ouroboros-consensus/docs/future-ledger-hd.md

```diff
@@ -0,0 +1,219 @@
+> `(re)applyBlockOpts`.
+
+### Stake snapshots
+
+The snapshots are rotated by the `snapTransition` rule. This is the most
+complicated of the three because it involves accessing the unified map and the
+`IncrementalStake` in their entirety in order to fold them. We do these two
+steps:
+
+```
+step1 = (dom activeDelegs ◁ credStake) ∪ (dom activeDelegs ◁ ptrStake)
+step2 =  aggregate (dom activeDelegs ◁ rewards) step1
+```
+
+Where 
+- `activeDelegs` comes from the delegations in the unified map,
+- `rewards` come from the unified map,
+- `credStake` and `ptrStake` come from the incremental stake
+
+However, there is an important note here. The `ssStakeMark` is only used to be
+put in `ssStakeSet` on the next snapshot rotation, then `ssStakeSet` is only
+used to be put in `ssStakeGo`, and `ssStakeGo` is only used to prepare the
+`PulsingRewUpdate`. This is done in `PulsingReward.startStep` and it also uses
+the `rewards` map of the unified map. Assuming the reward calculation is done
+outside of the ledger rules, the snapshot is not really needed by the ledger.
+
+The `ssStakeMarkPoolDistr` field is used to be put in `nesPd` on the
+`NewEpochState` (by `NEWEPOCH`) which later will be used to provide ledger views
+and calculate leader schedules. Note this is purely a Protocol concern and thus
+probably a Consensus concern.
+
+However it seems that the `ssPoolParams` are in fact modified by the Ledger.
+This should not be very problematic as we would know in andvance which pools are
+updating their params, and we could replicate the schema above for this map.
+
+## Reward computation and the ADA pots
+
+The way the rewards computation happens now (see `PulsingRewards` and
+`Shelley.Rules.Rupd` and `Shelley.Rules.Tick`) is that on each ticking we pulse
+a chunk of the rewards update so that when we reach the epoch boundary we want
+to have pulsed through the whole `ssStakeGo` snapshot that was used when
+creating the pulser.
+
+In the end, the reward computation has to produce a `RewardUpdate`:
+```haskell
+data RewardUpdate c = RewardUpdate
+  { deltaT :: !DeltaCoin
+  , deltaR :: !DeltaCoin
+  , rs :: !(Map (Credential 'Staking c) (Set (Reward c)))
+  , deltaF :: !DeltaCoin
+  , nonMyopic :: !(NonMyopic c)
+  }
+```
+
+where we will use the `delta*` fields to update the treasury, reserves and fee
+pot on the `NEWEPOCH` rule. We also use the `rs` field when calling
+`updateRewards` which happens also on `NEWEPOCH`.
+
+
+If Consensus can compute the `RewardUpdate` (possibly on a separate thread that
+traverses the map at its own pace?) then we can provide the `RewardUpdate` to
+the `NEWEPOCH` rule so that the pots can be updated.
+
+> **_PLAN:_** Consensus will compute the rewards outside of the Ledger and will
+> provide the `RewardUpdate` to the ledger so that it can modify the pots and
+> also produce diffs for the unified map's `Rewards` field. Ledger will not
+> compute the `RewardUpdate` thus the pulser becomes dead code.
```

Comment:
I think I misworded the sentence. I was meaning:

> Consensus will compute the rewards outside of the Ledger and will
> provide the `RewardUpdate` to the Ledger so that the Ledger can modify the pots and
> also the Ledger can produce diffs for the unified map's `Rewards` field based on the withdrawals happening in this block

Reply:
Sorry if I am being overly picky! This doesn't sound right to me:

> [Consensus] will provide the RewardUpdate to the Ledger

and also this:

> the Ledger can produce diffs for the unified map's Rewards field

---

How about something like this:

> **_PLAN:_** Consensus will compute the rewards outside of the Ledger and will
> provide the _parts_ of the `RewardUpdate` to the ledger. In particular, on the epoch
> boundary, it will provide `deltaT`, `deltaR`, and `deltaF`.
> Ledger will not compute the `RewardUpdate` thus the pulser becomes dead code.
>
> Moreover, the ledger will provide a function `f :: Block -> Set (StakeCredential c)`
> so that Consensus can supply Ledger with only the part of the `UMap` (the unified map)
> that it requires. In particular, Consensus will have to apply the
> `rs :: !(Map (Credential 'Staking c) (Set (Reward c)))` field of the reward update
> to the unified map on the epoch boundary, _prior_ to computing this view for the ledger
> (to ensure that reward withdrawals are correct within the given block).